### DISCUSSION TITLE
Java 1일차 [서희, 재찬]

### DISCUSSION THREAD

**Discussion by LireEruel** [Shortudy-10th/tech-for-developer#19] [질의 응답]
- 자바 소스 파일의 컴파일 & 실행 과정에 대해 설명해주세요.
- Call By Value / Call By Reference에 대해 설명해주세요.
- 자바의 자료형에 대해 아는 대로 설명해주세요.
- Java Collection에 대해서 설명해주세요.
- JAVA에는 접근 제어자가 4가지 있습니다. 각각의 이름을 말씀해 주시고, 각각 제어자의 차이점을 설명해 주세요.

**Comment by kjc6735** [comment]
1.자바 소스 파일의 컴파일 & 실행 과정에 대해 설명해주세요.
자바 컴파일러를 통해서 .class파일로 변환한다. 이 후 실행 할 때, 필요한 클래스 파일들을 로드한다. 인터프리터를 통해 프로그램이 동작한다.

2. Call By Value / Call By Reference에 대해 설명해주세요.
Call By Value는 값을 복사하여 보내는 것이므로 값을 변경해도 원본에는 변경이 없다.  Call by Reference는 주소 값을 넘겨서 수정하면 원본에 영향이 간다. 

 
 3. 자바의 자료형에 대해 아는 대로 설명해주세요.
 자바의 자료형에는 primitive type, reference type이 있다. primitive type는 스택에 저장되며, 파라미터로 보낼 때, call by value로 넘어간다. 또한 null값을 가질 수 없다. 종류에는 byte, char, short, boolean ,int, long, double, float가 있다. 레퍼런스타입은 오브젝트를 상속하고, 힙메모리에 올라간다. 
 
 4. Java Collection에 대해서 설명해주세요.
컬렉션은 컬렉션스와 Map이 있다. 둘을 합쳐서 컬렉션 프레임워크라고 부른다., 컬렉션스는   list, set이 있으며, 리스트는 순서가 보장되며, 셋은 순서가 보장되지 않는다. 맵은 키와 벨류로 구성되어 있으며, 순서가 보장되어 있지 않다.

5. JAVA에는 접근 제어자가 4가지 있습니다. 각각의 이름을 말씀해 주시고, 각각 제어자의 차이점을 설명해 주세요.
자바의 접근 제어자는 public, default, protected, private 가 있다. 퍼블릭에서는 어디서든 접근이 가능하다. protected는 자식 클래스에서 접근이 가능하며, default는 패키지 내에서 접근이 가능하다. private는 같은 클래스 내에서 접근이 가능하다.

**Reply by LireEruel** [reply]
Call by Value와 Reference 각각 언제 동작하는지 알려주세요

**Comment by LireEruel** [comment]
1. Java 컴파일러가 java  파일을 클래스파일로 변환해준다. 그 후 실행될 때는 class파일이 사용된다.
2. Call by Value는 값을 복사하여 넘겨주는 것이고 Call by Reference는 객체의 참조 값을 전달하는 것이다. 자바는 대부분 Call by Value이다. 
3. 자바의 자료형에는 primitive 타입과 reference 타입이 있다. primitive 타입에는 boolean, int, float, char, long, double이 있고 reference에는 Object, String 등이 있다. reference 타입은 상속을 해줄 수 있으며 최상위 객체는 Object이다. primitive 타입을 가진 멤버 변수는 클래스 생성시 초기화가 되지만 reference 타입은 초기화가 되지 않는다.
4. Java Collection Framework는 자바의 모든 컬렉션 클래그와 인터페이스를 포함하는 개념입니다. Collection 인터페이스는 List, Set과 Queue 3가지의 상위 인터페이스로 분류할 수 있다. Collections를 사용하면 3가지의 장점이 있습니다. 1. 일괄된 API, 2. 프로그래밍 노력 감소, 3. 프로그램 속도 및 품질 향상
5. private public protected default. public : 어디서든 접근이 가능하다. protected 자식 클래스에서 접근이 가능하다. default 같은 패키지 내에서 접근 가능하다. private 같은 클래스 내에서 접근 가능하다.

**Reply by kjc6735** [reply]
그럼 객체를(Object)를 파라미터로 넘긴다면 Call by value와 Call by Refernce 중 어떤 방식으로 동작하나요? 그리고 그렇게 생각한 이유를 말씀해주세요.

**Reply by LireEruel** [reply]
JAVA에서 메소드 호출을 할 때 인자값은 항상 Call by value 로 동작합니다. 객체를 전달 할 때에는 메모리 주소값이 value이기 때문에 객체를 수정하면 원본 객체에 영향이 끼칩니다. 하지만 call by reference와의 차이점은 인자값에 새로운 객체를 할당 했을 때 Call by Value는 원본 객체에 영향을 끼치지 않지만 Call by refernce는 영향을 미친다는 점 입니다.
